File: ts/src/optimization/effectiveCombination.ts
const LOSE_SCORES = [-10, -9, -8, -7, -6, -5, -4, -3, -2, -1];
const DRAW_SCORES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
const WIN_SCORES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

const cartesianProduct: () => number[][] = () => {
  const result = [];
  for (const lose of LOSE_SCORES) {
    for (const draw of DRAW_SCORES) {
      for (const win of WIN_SCORES) {
        result.push([lose, draw, win]);
      }
    }
  }
  return result;
};

export const effectiveCombination: () => number[][] = () => {
  // Filter out the combinations that are not effective:
  // 1. winScore <= drawScore
  // 2. loseScore, drawScore, winScore have common factor
  const rawProduct = cartesianProduct();
  const first = rawProduct.filter(([_lose, draw, win]) => win > draw);
  const second = first.filter((arr) => {
    const [a, b, c] = arr.map(Math.abs);
    const min = Math.min(a, b, c);
    for (let factor = 2; factor <= min; factor++) {
      if (a % factor === 0 && b % factor === 0 && c % factor === 0) return false;
    }
    return true;
  });
  return second;
};

File: java/src/main/java/com/cc/optimization/EffectiveCombination.java
package com.cc.optimization;

import java.util.ArrayList;
import java.util.List;

public class EffectiveCombination {
    static final int[] LOSE_SCORES = {-10, -9, -8, -7, -6, -5, -4, -3, -2, -1};
    static final int[] DRAW_SCORES = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    static final int[] WIN_SCORES = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    private static List<int[]> cartesianProduct() {
        List<int[]> result = new ArrayList<>();
        for (int lose : LOSE_SCORES) {
            for (int draw : DRAW_SCORES) {
                for (int win : WIN_SCORES) {
                    result.add(new int[]{lose, draw, win});
                }
            }
        }
        return result;
    }

    public static List<int[]> getAll() {
        // Filter out the combinations that are not effective:
        // 1. winScore <= drawScore
        // 2. loseScore, drawScore, winScore have common factor
        List<int[]> rawProduct = cartesianProduct();
        List<int[]> first = new ArrayList<>();
        for (int[] arr : rawProduct) {
            int lose = arr[0], draw = arr[1], win = arr[2];
            if (win > draw) {
                first.add(arr);
            }
        }

        List<int[]> second = new ArrayList<>();
        for (int[] arr : first) {
            int a = Math.abs(arr[0]), b = Math.abs(arr[1]), c = Math.abs(arr[2]);
            int min = Math.min(Math.min(a, b), c);
            boolean hasCommonFactor = false;
            for (int factor = 2; factor <= min; factor++) {
                if (a % factor == 0 && b % factor == 0 && c % factor == 0) {
                    hasCommonFactor = true;
                    break;
                }
            }
            if (!hasCommonFactor) {
                second.add(arr);
            }
        }

        return second;
    }
}
